server/src/main/java/org/elasticsearch/script/field/MapBackedMetadata.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.script.field;
+
+import org.elasticsearch.common.util.Maps;
+
+import java.util.Map;
+import java.util.Objects;
+
+/** Metadata storage backed by a Map for compatibility with the ctx Map used by ingest and update scripts. */
+public class MapBackedMetadata {
+    public static final String INDEX = "_index";
+    public static final String ID = "_id";
+    public static final String VERSION = "_version";
+    public static final String ROUTING = "_routing";
+    public static final String SOURCE = "_source";
+
+    protected Map<String, Object> map;
+
+    public MapBackedMetadata(Map<String, Object> map) {
+        this.map = map;
+    }
+
+    public MapBackedMetadata(int size) {
+        this.map = Maps.newMapWithExpectedSize(size);
+    }
+
+    public String getIndex() {
+        return getString(INDEX);
+    }
+
+    public MapBackedMetadata setIndex(String index) {
+        return set(INDEX, index);
+    }
+
+    public String getId() {
+        return getString(ID);
+    }
+
+    public MapBackedMetadata setId(String id) {
+        return set(ID, id);
+    }
+
+    public Long getVersion() {
+        Object obj = getRawVersion();
+        if (obj == null) {
+            return null;
+        } else if (obj instanceof Number number) {
+            long version = number.longValue();
+            if (number.doubleValue() != version) {
+                // did we round?
+                throw new IllegalArgumentException(
+                    "version may only be set to an int or a long but was [" + number + "] with type [" + obj.getClass().getName() + "]"
+                );
+            }
+            return version;
+        }
+        throw new IllegalArgumentException(
+            "version may only be set to an int or a long but was [" + obj + "] with type [" + obj.getClass().getName() + "]"
```

Comment:
Seeing this exception on a get is confusing. We should perhaps validate the ctx map when it is passed into the ctor?

Reply:
The script still has access to ctx and may change it there.  Unfortunately, that means we cannot trust the contents of the map at any point.

Reply:
Yup, we'd have to override `merge`, `put`, `putAll`, `putIfAbsent`, `replace(k,v)`, `replace(k,v,v)` and `replaceAll`.

Would that be a backwards incompatible change?  It would be rejected after the script runs, but during script execution the map can hold anything as long as it's fixed up before script termination.

Reply:
Would the wrapped map block/validate all changes, or just do this same type of validation for the [known metadata fields](https://github.com/elastic/elasticsearch/pull/87309/files/502aac29619e4c4021b69085d584deb1b9a30e7c#diff-6b5913f6e06049c3e3c48eb973883a79554093b08b481006d8af106a92a8680cR18-R22)?

Reply:
It would block writing to unknown keys. ie the only valid keys are the known metadata fields (the list you linked isn't all of them, it depends on the script context, but those I think are the ones for ingest, or most of them).

Reply:
Actually, on further inspection of IngestDocument, I think we can do this as I originally suggested @stu-elastic . The sourceAndMetadata map is what we need to wrap. On put, if the key is one of the known metadata fields, we validate the value (this would get validated later anyways when we try to index, this is just moving it more upfront). Unknown keys would just be delegated without validation because those are actually document keys.

Reply:
To help illustrate the BWC issue, the following valid script would become invalid if we validated the map.
```
ctx._version = 'foo';
...
ctx._version = (ctx._version == 'foo') ? 10: 20;
```

Reply:
IIUC, we're only talking about validating the known metadata fields and the user-facing change would be that they're validated every time they're set rather than only being validated at the end of the script (or, when they're read from the context as part of document storage).

That feels OK to me. If we only validate on `getVersion` (etc) then there's a cost to the user that they cannot easily trace where the version was set. They know that something in their script set the version to a string, but that could potentially happen in more than 1 place.

So, yes, there's the possibility that it might break something for someone, but that's true of many bug fixes. This feels like a reasonable feature to validate on set rather than validate on first use. I agree it's borderline, and someone might have a compelling argument for why we should consider it "breaking" rather than "fixing", but I think I'm OK with it.